incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -38,34 +45,32 @@ type ConfigReconciler struct {
 // Reconcile sets up some basic variable and logs the Spec.
 // TODO: Updates the comment above when adding more logic to the Reconcile method.
 func (r *ConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO: Surface the error more prominently and add a validating admission controller to prevent
-	// the problem in the first place.
-	if req.NamespacedName.Name != api.HNCConfigSingleton {
-		r.Log.Error(nil, "Singleton name is wrong. It should be 'config'.")
+	ctx := context.Background()
+
+	// Validate the singleton name.
+	// TODO: Add a validating admission controller to prevent the problem in the first place.
+	if err := r.validateSingletonName(ctx, req.NamespacedName.Name); err != nil {
+		r.Log.Error(err, "Singleton name validation failed.")
 		return ctrl.Result{}, nil
 	}
```

Comment:
I'd put the error message in the "real" singleton - after all, that's where users will go to look for their problems. So this should really be something like:

`r.validateSingletonName(ctx, inst, req.NamespacedName);`

Note that this doesn't return an error - it just adds the condition to the _real_ singleton. You could _also_ add a condition to the "bad" config as well, if you like, but note that if you can't write back to it for any reason, you should probably just log that error and move on to try to write to the "real" one.

Reply:
It doesn't look like this will work (also tested on the real cluster). When we do reconcile, we always update Status (i.e., Condition in the Status) to the latest one (feel free to correct me if this is not the case). If we write the error message to the correct singleton (one with name 'config'), the workflow will be like:
 
 1. Write the error condition to the correct singleton
 2.  Singleton change above will trigger a new reconciliation for the correct singleton. The request for the reconciliation will have the correct name so the Status of the correct singleton will be updated again and the error message will be removed.
 3.  Singleton change above will trigger a new reconciliation for the correct singleton. Nothing changed at this point, no more subsequent reconciliations triggered by invalid name.  

When user check the status of the correct singleton, it will be after step 3 when all the reconciliation chain finishes. At that point, the error message is already removed.  

Feel free to let me know if I miss anything. 

------------------------------------------------------------

Regarding to `validateSingletonName` not returning an error... I don't think it has to return an error, but I feel it should return some signal to indicate if the name is wrong or not. If the name is wrong, we don't need to execute rest of steps in the `Reconcile` (e.g., creating/updating object reconciliers). IIUC, we only consider the `Spec` in the correct singleton, not the wrong one.

Reply:
Hi @adrianludwin 

Friendly ping in case you miss this reply. There are some issues about writing the error message to the correct object as I described above. Feel free to take a look. Thank you!

Reply:
Ok I think this is fine for now. I'll approve.